keras_applications/resnet_v2.py

```diff
@@ -0,0 +1,37 @@
+"""ResNetV2 models for Keras.
+
+# Reference paper
+
+- [Aggregated Residual Transformations for Deep Neural Networks]
+  (https://arxiv.org/abs/1611.05431) (CVPR 2017)
+
+# Reference implementations
+
+- [TensorNets]
+  (https://github.com/taehoonlee/tensornets/blob/master/tensornets/resnets.py)
+- [Torch ResNetV2]
+  (https://github.com/facebook/fb.resnet.torch/blob/master/models/preresnet.lua)
+
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from . import imagenet_utils
+from .imagenet_utils import decode_predictions
+from .resnet_common import ResNet50V2
+from .resnet_common import ResNet101V2
+from .resnet_common import ResNet152V2
+
+
+def preprocess_input(x, **kwargs):
+    """Preprocesses a numpy array encoding a batch of images.
+
+    # Arguments
+        x: a 4D numpy array consists of RGB values within [0, 255].
+        data_format: data format of the image tensor.
+
+    # Returns
+        Preprocessed array.
+    """
+    return imagenet_utils.preprocess_input(x, mode='tf', **kwargs)
```

Comment:
I feel like it is somewhat error-prone that ResNet50,  ResNeXt50, and ResNet50V2 use 3 different preprocessing methods. Is there anything we could do about that?

Reply:
@fchollet, I agree that several users may be confused. However, it is not easy to revise because the weights are originated from different researchers using different libraries. I think that one of the possible ways is to insert addition and division operations into model definitions, but it is not consistent with the existing model definition styles.

Reply:
I think that having the preprocessing in the model definition make a lot of sense and is a good idea. Models in `keras.application` cannot be used without the appropriate preprocessing, even when fine-tuning, when `weights` is not `None`. 

I would be in favor of adding the preprocessing in the model when `weights='imagenet'` but it would break the API and change the existing behavior. Can we maybe add a new argument when creating the model? `preprocessing='imagenet' (or None)`? 

I think it adds a lot of unnecessary thinking to users when trying to use pre-trained models. (the number of issues and stackoverflow questions about pre-processing when using `keras.applications` is staggering). It also comes from the functions being all called `preprocess_input` ,so users who don't know the details of the preprocessing of the different models think that the same `preprocessing_input` will cover all models.

There are also performance issues to consider since the preprocessing will happen on the GPU. Since this is close to an element-wise operation in most cases, this should be fine since GPUs are good at this.

Thoughts?

Reply:
@gabrieldemarmiesse, Thank you for the comment. The `preprocessing='imagenet'` idea looks good to me. However, it is not easy to change the styles because users who use the current styles (`model.predict(preprocess_input(x))`) must be confused if a preprocessing is included in a model definition. If such is the case, the current styles will perform the preprocessing twice and result in wrong outputs. I think that the style change is a matter of choice.

Reply:
I see what you mean. It's true that it can be a source of confusion for users. I guess we can let @fchollet decide if the API needs a modification.

Reply:
@fchollet, How about the following two styles?
- A preprocessing is decouple from a model definition (the current): Several users may be confused because ResNet, ResNetV2, and ResNeXt have similar names. However, in my opinion, there will be no problem for users who are familiar with the current style.
- A preprocessing is defined within a model definition (an alternative): It is reasonable because the certain structure with its pre-trained weights must be coupled with specific preprocessing. However, users who use the current styles (`model.predict(preprocess_input(x))`) must be confused if a preprocessing is included in a model definition. Something complimentary such as a user warning needs to be proposed together.

Reply:
@taehoonlee How to fire the warning? In which situations/API calls?

Reply:
@gabrieldemarmiesse, I think that one of the simple ideas is to put a warning like "You must feed a [0, 1] scaled image without any preprocessing since Keras 2.x.x" into every model definition. I have no elegant idea.

Reply:
If we depreciate the numpy version of `preprocess_input`, we would throw a warning whenever it is called. I'm thinking of a message like:
`preprocess_input is depreciated, please use 'preprocess=' imagenet'' when calling the model`. 

We would solve the issues of:

* Users can' t use the preprocessing twice since they would see the warning.
* We could migrate very slowly towards the argument `preprocessing` defaulting to the same value as `weights` which is the behaviour that a new user would expect when using `keras.application`. Thus reducing the cognitive workload and verbosity.

What are your thoughts on this?

Reply:
This new suggested API doesn't sound transparent to me. Why not let the pre-processing functions be included in the pre-processing module and define a number of preprocessing methods based on different papers, and provide a mapping, so that users can use the right one when using pre-trained weights, and choose anyone when training from scratch ? A bit like Tensorflow Slim does it:

https://github.com/tensorflow/models/tree/master/research/slim/preprocessing

Something like this:

```python
from keras.preprocessing.image import preprocess_input_for_cnn

# Load image 
...

input_to_network  =  preprocess_input_for_cnn(image, method="inception")
input_to_network  =  preprocess_input_for_cnn(image, method="alexnet")
input_to_network  =  preprocess_input_for_cnn(image, method="vgg")



```

Reply:
@gabrieldemarmiesse, Yes, I agree with your basic principles. However, it would be perfect if there was a way to force users not to run preprocessing twice, instead of giving a warning message. I have no idea, thus frankly speaking, I think the current form is better from the perspective of users who are familiar with the current one.

Reply:
> However, it would be perfect if there was a way to force users not to run preprocessing twice, instead of giving a warning message.

I totally agree. If we use a global variable, it would work, but I don't really like global variables. I'm not even sure it's going to work well with multithreading/multiprocessing.

Generally speaking, I would really like a way to avoid silent preprocessing errors (they are VERY common for many users), after that, which API we choose doesn't really matter to me.

Reply:
@rohit-gupta, I think that is overkill. It must result in performance degradation.

Reply:
I may be wrong, but if the preprocessing is declared in the graph, a graph optimizer (let's say XLA for example) can fold it in the first convolution, making the preprocessing time drop to 0. I may be wrong but I think it's possible. Not that it's going to be significant compared to the execution time of the full graph. But speedups are always nice.